products/llbuildSwift/BuildDBBindings.swift

```diff
@@ -0,0 +1,345 @@
+    public let dependencies: [KeyID]
+    
+    fileprivate init(result: llb_database_result_t) {
+        self.result = result
+        value = Data(bytes: result.value.data, count: Int(result.value.length))
+        signature = result.signature
+        computedAt = result.computed_at
+        builtAt = result.built_at
+        
+        dependencies = UnsafeBufferPointer(start: result.dependencies, count: Int(result.dependencies_count)).map { KeyID($0) }
+    }
+    
+    deinit {
+        withUnsafeMutablePointer(to: &result) { ptr in
+            llb_database_destroy_result(ptr)
+        }
+    }
+    
+    /// Fetch result information about all dependencies via the given database
+    public func fetchDependencies(database: Database.BuildDB) throws -> [RuleResult] {
+        var result = [RuleResult]()
+        
+        for dep in dependencies {
+            guard let res = try database.lookupRuleResult(keyID: dep) else {
+                continue
+            }
+            result.append(res)
+        }
+        
+        return result
+    }
+}
+
+extension RuleResult: CustomReflectable {
+    public var customMirror: Mirror {
+        // We exclude the result pointer from the reflection
+        let properties: DictionaryLiteral<String, Any> = [
+            "signature": signature,
+            "computedAt": computedAt,
+            "builtAt": builtAt,
+            "dependencies": dependencies,
+            "value": value
+        ]
+        
+        return Mirror(RuleResult.self, children: properties, displayStyle: .class, ancestorRepresentation: .generated)
+    }
+}
+
+/// Private class for easier handling of out-parameters
+private class MutableStringPointer {
+    var ptr = llb_data_t()
+    init() { }
+    
+    deinit {
+        ptr.data?.deallocate()
+    }
+    
+    var msg: String? {
+        guard ptr.data != nil else { return nil }
+        return stringFromData(ptr)
+    }
+}
+
+/// Database object that defines a connection to a llbuild database
+public final class Database {
+    
+    /// Errors that can happen when opening the database or performing operations on it
+    public enum Error: Swift.Error {
+        /// If the system can't open the database, this error is thrown at init
+        case couldNotOpenDB(String)
+        /// If an operation on the database fails, this error is thrown
+        case operationDidFail(String)
+        /// If the database didn't provide an error but the operation still failed, the unknownError is thrown
+        case unknownError
+    }
+    
+    public private(set) var delegate: BuildDBDelegate
+    /// The opaque pointer to the database object
+    private var _database: OpaquePointer? = nil
+    
+    /// Initializes the build database at a given path
+    /// If the database at this path doesn't exist, it will created
+    /// If the clientSchemaVersion is different to the one in the database at this path, its content will be automatically erased!
```

Comment:
This is fairly dangerous, and probably not what we want. We should probably make it so that this doesn't happen, and is instead an error.

Reply:
But that should only be done in this layer, not in the underlying SQLBuildDB right?